datafusion/physical-expr/src/regex_expressions.rs

```diff
@@ -78,6 +79,82 @@ pub fn regexp_match<T: OffsetSizeTrait>(args: &[ArrayRef]) -> Result<ArrayRef> {
     }
 }
 
+/// TODO: Remove this once it is included in arrow-rs new release.
+fn _regexp_match<OffsetSize: OffsetSizeTrait>(
+    array: &GenericStringArray<OffsetSize>,
+    regex_array: &GenericStringArray<OffsetSize>,
+    flags_array: Option<&GenericStringArray<OffsetSize>>,
+) -> std::result::Result<ArrayRef, ArrowError> {
+    let mut patterns: std::collections::HashMap<String, Regex> =
```

Comment:
This is awesome -- I think we can do better still (this will still recompile each regex per batch I think -- we can do it once per plan)

Reply:
Yea, it is per batch (not per row). Definitely if we do the regex compilation per query plan, it will be better. 🚀

Reply:
Btw, to clarify it if the long PR description is overlooked by reviewers. This hash map is already in the arrow-rs kernel. The actual fix here is to avoid expensive cloning of `Regex` per row. I described the reason in the description.

Reply:
@alamb are you thinking to have some static shared memory pool per query to make static values to be calculated only once per query? we can prob do that with `lazy_static` and `std::sync::Once`

Reply:
I think @alamb means something like Pre-compiled / Prepared ScalarUDFs (https://github.com/apache/arrow-datafusion/issues/8051).